### ISSUE TITLE
Netrw special crashes in vim 9.0 1-815 

### ISSUE THREAD

**Issue by fvdnabee** [vim/vim#11457]
### Steps to reproduce

1. Activate netrw
2. Hit 'x' on any file
3. Hit 'x' again on any file: SIGSEGV

### Expected behaviour

netrw opens gnome text editor (if file was text file).

Invoking xdg-open myself from the CLI opens the gnome text editor.

### Version of Vim

9.0 1-815

### Environment

Operation system: ubuntu jammy
Terminal: tested in alacritty and gnome-terminal
$TERM: xterm-256color
shell: tested in zsh and bash
Using wayland ($XDG_SESSION_TYPE = wayland).

### Logs and stack traces

```shell
❯ gdb --args ~/GIT/vim/src/vim -D -U NONE -u ~/test-vimrc .

#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737319774848) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737319774848) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737319774848, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff6c2f476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff6c157f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff6c766f6 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff6dc8b8c "%s\n") at ../sysdeps/posix/libc_fatal.c:155
#6  0x00007ffff6c8dd7c in malloc_printerr (str=str@entry=0x7ffff6dcb710 "free(): double free detected in tcache 2") at ./malloc/malloc.c:5664
#7  0x00007ffff6c9012b in _int_free (av=0x7ffff6e06c80 <main_arena>, p=0x555555e4a680, have_lock=0) at ./malloc/malloc.c:4473
#8  0x00007ffff6c924d3 in __GI___libc_free (mem=<optimized out>) at ./malloc/malloc.c:3391
#9  0x00005555555f1e06 in vim_free (x=0x555555e4a690) at alloc.c:615
#10 0x000055555568076b in do_bang (addr_count=0, eap=0x7fffffff9a40, forceit=0, do_in=1, do_out=1) at ex_cmds.c:965
#11 0x000055555569a129 in ex_bang (eap=0x7fffffff9a40) at ex_docmd.c:8169
#12 0x000055555568edc7 in do_one_cmd (cmdlinep=0x7fffffff9c70, flags=3, cstack=0x7fffffff9d50, fgetline=0x55555587cd6d <get_func_line>, cookie=0x555555e52c90) at ex_docmd.c:2578
#13 0x000055555568bc79 in do_cmdline (cmdline=0x555555e4be20 "sil !xdg-open '/home/fvdnabee/GIT/vim-rhubarb/README.markdown'>/dev/null 2>&1&", fgetline=0x55555587cd6d <get_func_line>, cookie=0x555555e52c90, flags=3) at ex_docmd.c:990
#14 0x000055555565e2f6 in ex_execute (eap=0x7fffffffa530) at eval.c:6957
#15 0x000055555568edc7 in do_one_cmd (cmdlinep=0x7fffffffa760, flags=7, cstack=0x7fffffffa840, fgetline=0x55555587cd6d <get_func_line>, cookie=0x555555e52c90) at ex_docmd.c:2578
#16 0x000055555568bc79 in do_cmdline (cmdline=0x0, fgetline=0x55555587cd6d <get_func_line>, cookie=0x555555e52c90, flags=7) at ex_docmd.c:990
#17 0x0000555555875853 in call_user_func (fp=0x555555e33ce0, argcount=1, argvars=0x7fffffffb6e0, rettv=0x7fffffffb890, funcexe=0x7fffffffb8a0, selfdict=0x0) at userfunc.c:2945
#18 0x0000555555875ee5 in call_user_func_check (fp=0x555555e33ce0, argcount=1, argvars=0x7fffffffb6e0, rettv=0x7fffffffb890, funcexe=0x7fffffffb8a0, selfdict=0x0) at userfunc.c:3107
#19 0x0000555555876fb4 in call_func (funcname=0x555555e4d870 "\200\375R23_NetrwExe", len=-1, rettv=0x7fffffffb890, argcount_in=1, argvars_in=0x7fffffffb6e0, funcexe=0x7fffffffb8a0) at userfunc.c:3663
#20 0x00005555558730f7 in get_func_tv (name=0x555555e4d870 "\200\375R23_NetrwExe", len=-1, rettv=0x7fffffffb890, arg=0x7fffffffb940, evalarg=0x7fffffffb9e0, funcexe=0x7fffffffb8a0) at userfunc.c:1841
#21 0x000055555587bcdd in ex_call_inner
     (eap=0x7fffffffbb30, name=0x555555e4d870 "\200\375R23_NetrwExe", arg=0x7fffffffb940, startarg=0x555555e2fc02 "(\"sil !xdg-open \".s:ShellEscape(fname,1).redir.'&')", funcexe_init=0x7fffffffb990, evalarg=0x7fffffffb9e0) at userfunc.c:5647
#22 0x000055555587c939 in ex_call (eap=0x7fffffffbb30) at userfunc.c:5971
#23 0x000055555568edc7 in do_one_cmd (cmdlinep=0x7fffffffbd60, flags=7, cstack=0x7fffffffbe40, fgetline=0x55555587cd6d <get_func_line>, cookie=0x555555d566a0) at ex_docmd.c:2578
#24 0x000055555568bc79 in do_cmdline (cmdline=0x0, fgetline=0x55555587cd6d <get_func_line>, cookie=0x555555d566a0, flags=7) at ex_docmd.c:990
#25 0x0000555555875853 in call_user_func (fp=0x555555dcde10, argcount=2, argvars=0x7fffffffcce0, rettv=0x7fffffffce90, funcexe=0x7fffffffcea0, selfdict=0x0) at userfunc.c:2945
#26 0x0000555555875ee5 in call_user_func_check (fp=0x555555dcde10, argcount=2, argvars=0x7fffffffcce0, rettv=0x7fffffffce90, funcexe=0x7fffffffcea0, selfdict=0x0) at userfunc.c:3107
#27 0x0000555555876fb4 in call_func (funcname=0x555555a478a0 "netrw#BrowseX", len=-1, rettv=0x7fffffffce90, argcount_in=2, argvars_in=0x7fffffffcce0, funcexe=0x7fffffffcea0) at userfunc.c:3663
#28 0x00005555558730f7 in get_func_tv (name=0x555555a478a0 "netrw#BrowseX", len=-1, rettv=0x7fffffffce90, arg=0x7fffffffcf40, evalarg=0x7fffffffcfe0, funcexe=0x7fffffffcea0) at userfunc.c:1841
#29 0x000055555587bcdd in ex_call_inner
     (eap=0x7fffffffd130, name=0x555555a478a0 "netrw#BrowseX", arg=0x7fffffffcf40, startarg=0x555555d4ea92 "(<SNR>23_NetrwBrowseChgDir(1,<SNR>23_NetrwGetWord(),0),0)\"", funcexe_init=0x7fffffffcf90, evalarg=0x7fffffffcfe0) at userfunc.c:5647
#30 0x000055555587c939 in ex_call (eap=0x7fffffffd130) at userfunc.c:5971
#31 0x000055555568edc7 in do_one_cmd (cmdlinep=0x7fffffffd360, flags=0, cstack=0x7fffffffd440, fgetline=0x5555556a6e78 <getexline>, cookie=0x0) at ex_docmd.c:2578
#32 0x000055555568bc79 in do_cmdline (cmdline=0x0, fgetline=0x5555556a6e78 <getexline>, cookie=0x0, flags=0) at ex_docmd.c:990
#33 0x000055555573e39e in nv_colon (cap=0x7fffffffdb30) at normal.c:3183
#34 0x000055555573a1b3 in normal_cmd (oap=0x7fffffffdbc0, toplevel=1) at normal.c:920
#35 0x0000555555943d8b in main_loop (cmdwin=0, noexmode=0) at main.c:1524
#36 0x0000555555943031 in vim_main2 () at main.c:886
#37 0x00005555559426a4 in main (argc=7, argv=0x7fffffffddf8) at main.c:432
```

```
❯ cat ~/test-vimrc 
set nocompatible

syntax on
filetype plugin indent on
```

<details>
<summary>vim version</summary>
Latest patch version compiled from source, with debug enabled.
<pre>
❯ ~/GIT/vim/src/vim --version
VIM - Vi IMproved 9.0 (2022 Jun 28, compiled Oct 28 2022 10:49:50)
Included patches: 1-815
Compiled by fvdnabee@oqubu
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl               +file_in_path      +mouse_urxvt       -tag_any_white
+arabic            +find_in_path      +mouse_xterm       +tcl/dyn
+autocmd           +float             +multi_byte        +termguicolors
+autochdir         +folding           +multi_lang        +terminal
-autoservername    -footer            -mzscheme          +terminfo
+balloon_eval      +fork()            +netbeans_intg     +termresponse
+balloon_eval_term +gettext           +num64             +textobjects
+browse            -hangul_input      +packages          +textprop
++builtin_terms    +iconv             +path_extra        +timers
+byte_offset       +insert_expand     +perl/dyn          +title
+channel           +ipv6              +persistent_undo   +toolbar
+cindent           +job               +popupwin          +user_commands
+clientserver      +jumplist          +postscript        +vartabs
+clipboard         +keymap            +printer           +vertsplit
+cmdline_compl     +lambda            +profile           +vim9script
+cmdline_hist      +langmap           -python            +viminfo
+cmdline_info      +libcall           +python3/dyn       +virtualedit
+comments          +linebreak         +quickfix          +visual
+conceal           +lispindent        +reltime           +visualextra
+cryptv            +listcmds          +rightleft         +vreplace
+cscope            +localmap          +ruby/dyn          +wildignore
+cursorbind        +lua/dyn           +scrollbind        +wildmenu
+cursorshape       +menu              +signs             +windows
+dialog_con_gui    +mksession         +smartindent       +writebackup
+diff              +modify_fname      -sodium            +X11
+digraphs          +mouse             -sound             -xfontset
+dnd               +mouseshape        +spell             +xim
-ebcdic            +mouse_dec         +startuptime       +xpm
+emacs_tags        -mouse_gpm         +statusline        +xsmp_interact
+eval              -mouse_jsbterm     -sun_workshop      +xterm_clipboard
+ex_extra          +mouse_netterm     +syntax            -xterm_save
+extra_search      +mouse_sgr         +tag_binary
-farsi             -mouse_sysmouse    -tag_old_static
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK -pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/x86_64-linux-gnu -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -ggdb3 -DDEBUG
Linking: gcc -ggdb3 -o vim -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE -lm -ltinfo -lselinux -Wl,-E -fstack-protector-strong -L/usr/local/lib -L/usr/lib/x86_64-linux-gnu/perl/5.34/CORE -lperl -ldl -lm -lpthread -lcrypt -L/usr/lib/x86_64-linux-gnu -ltclstub8.6 -ldl -lz -lpthread -lm
  DEBUG BUILD

</pre>
</details>

**Comment by k-takata** [comment]
I think this is a duplicate of #11424.

**Comment by fvdnabee** [comment]
> I think this is a duplicate of #11424.

Indeed, following @mgedmin's suggestion to revert a part of commit 9c50eeb40117413bf59a9da904c8d0921ed0a6e6, I no longer see a segfault:
```
diff --git a/src/ex_cmds.c b/src/ex_cmds.c
index 676ab704b..44333d514 100644
--- a/src/ex_cmds.c
+++ b/src/ex_cmds.c
@@ -929,8 +929,6 @@ do_bang(
            STRCAT(t, newcmd);
        if (ins_prevcmd)
            STRCAT(t, prevcmd);
-       else
-           vim_free(t);
        p = t + STRLEN(t);
        STRCAT(t, trailarg);
        vim_free(newcmd);
```
I don't know what the downsides of this revert are, but not segfaulting anymore is a pretty big plus :)

Closed in favor of #11424. Might re-open later if problem is not fixed after #11424 is resolved.

**Comment by fvdnabee** [comment]
Fixed after updating to 9.0 patch 1-820
